File: libraries/doctemplates/template/recordtypedef.lua
--------------------------------------------------------------------------------
--  Copyright (c) 2012 Sierra Wireless.
--  All rights reserved. This program and the accompanying materials
--  are made available under the terms of the Eclipse Public License v1.0
--  which accompanies this distribution, and is available at
--  http://www.eclipse.org/legal/epl-v10.html
-- 
--  Contributors:
--       Kevin KIN-FOO <kkinfoo@sierrawireless.com>
--           - initial API and implementation and initial documentation
--------------------------------------------------------------------------------
return [[#
<a id ="$(anchor(_recordtypedef))"></a>
# --
# -- Descriptions
# --
#if _recordtypedef.shortdescription then
	<p>$( format( _recordtypedef.shortdescription ) )</p>
#end
#if _recordtypedef.description then
	<p>$( format( _recordtypedef.description ) )</p>
#end
#--
#-- Describe usage
#--
#if _recordtypedef.metadata and _recordtypedef.metadata.usage then
	<h2>Usage examples</h2>
#	for _, usage in ipairs(_recordtypedef.metadata.usage) do
		<pre><code>$( securechevrons(usage.description) )</code></pre>
#	end
#end
# --
# -- Describe type fields
# --
#if not isempty( _recordtypedef.fields ) then
	<h3>Field(s)</h3>
#	for name, item in sortedpairs( _recordtypedef.fields ) do
		$( applytemplate(item) )
#	end
#end ]]

File: plugins/org.eclipse.koneki.ldt/src/org/eclipse/koneki/ldt/core/internal/ast/models/api/RecordTypeDef.java
/*******************************************************************************
 * Copyright (c) 2011 Sierra Wireless and others.
 * All rights reserved. This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v1.0
 * which accompanies this distribution, and is available at
 * http://www.eclipse.org/legal/epl-v10.html
 *
 * Contributors:
 *     Sierra Wireless - initial API and implementation
 *******************************************************************************/
package org.eclipse.koneki.ldt.core.internal.ast.models.api;

import java.util.HashMap;

import org.eclipse.dltk.ast.ASTVisitor;

/**
 * Use to define a kind of type 'record'.
 */
public class RecordTypeDef extends TypeDef {

	private String name;
	private String documentation;
	private HashMap<String, Item> fields = new HashMap<String, Item>();

	public RecordTypeDef() {

	}

	/**
	 * @see org.eclipse.dltk.ast.ASTNode#traverse(org.eclipse.dltk.ast.ASTVisitor)
	 */
	@Override
	public void traverse(ASTVisitor visitor) throws Exception {
		if (visitor.visit(this)) {
			// traverse fields
			for (Item field : fields.values()) {
				field.traverse(visitor);
			}
			visitor.endvisit(this);
		}
	}

	public String getName() {
		return name;
	}

	public String getDocumentation() {
		return documentation;
	}

	public HashMap<String, Item> getFields() {
		return fields;
	}

	public void addField(final String fieldName, final Item item) {
		fields.put(fieldName, item);
		item.setParent(this);
	}

	public void setName(String name) {
		this.name = name;
	}

	public void setDocumentation(String documentation) {
		this.documentation = documentation;
	}

	public String toReadableString() {
		return "#" + name; //$NON-NLS-1$
	}
}
